{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# starmap gene expression cell typing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import anndata as an\n",
    "bad_ch_pick = True\n",
    "sc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=80, facecolor='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set = sc.read_h5ad('./SourceData/gene_expression.h5ad')\n",
    "ephys_set = sc.read_h5ad('./SourceData/electrophysiology.h5ad')\n",
    "joint_set = sc.read_h5ad('./SourceData/gene_ephys_joint.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "stages = gene_set.obs['stage'].unique()\n",
    "marker_ = ['GATA4','MYH6','HCN4','ATP2A2','TNNT2','MYH7','VCL','RYR2']\n",
    "sc.pl.umap(gene_set, color=marker_,use_raw=False,save='CorrelatedUMAP.pdf',cmap='Reds_r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(vector_friendly=False)\n",
    "sc.pl.umap(gene_set, color=['cell type'],palette = 'Set1',save='gene_cluster_umap.pdf',size=250)\n",
    "sc.pl.umap(gene_set, color=['stage'], palette = 'rainbow',save='gene_stage_umap.pdf',size=250)\n",
    "\n",
    "\n",
    "sc.pl.umap(ephys_set, color=['cell type'],palette = 'Set2',save='ephys_cluster_umap.pdf',size=250)\n",
    "sc.pl.umap(ephys_set, color=['stage'], palette = 'rainbow',save='ephys_stage_umap.pdf',size=250)\n",
    "\n",
    "sc.pl.umap(joint_set, color=['cluster'],palette = 'spring',save='joint_cluster_umap.pdf',size=250)\n",
    "sc.pl.umap(joint_set, color=['stage'], palette = 'rainbow',save='joint_stage_umap.pdf',size=250)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "This code was adapted from https://github.com/AllenInstitute/coupledAE-patchseq\n",
    "Credit to original authors\n",
    "'''\n",
    "from sklearn.metrics import confusion_matrix\n",
    "import seaborn as sn\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import matplotlib\n",
    "from scipy.optimize import linear_sum_assignment\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as pex\n",
    "\n",
    "def color_cm(cmap,NUM_COLORS,):\n",
    "    color = []\n",
    "    color_idx = 0\n",
    "    cm = pylab.get_cmap(cmap)\n",
    "    for i in range(NUM_COLORS):\n",
    "        color.append(matplotlib.colors.to_hex(cm(1. * i / NUM_COLORS)))  # color will now be an RGBA tuple\n",
    "    return color\n",
    "def contingency(a, b, unique_a, unique_b):\n",
    "    \"\"\"Populate contingency matrix. Rows and columns are not normalized in any way.\n",
    "    \n",
    "    Args:\n",
    "        a (np.array): labels\n",
    "        b (np.array): labels\n",
    "        unique_a (np.array): unique list of labels. Can have more entries than np.unique(a)\n",
    "        unique_b (np.array): unique list of labels. Can have more entries than np.unique(b)\n",
    "\n",
    "    Returns:\n",
    "        C (np.array): contingency matrix.\n",
    "    \"\"\"\n",
    "    assert a.shape == b.shape\n",
    "    C = np.zeros((np.size(unique_a), np.size(unique_b)))\n",
    "    for i, la in enumerate(unique_a):\n",
    "        for j, lb in enumerate(unique_b):\n",
    "            C[i, j] = np.sum(np.logical_and(a == la, b == lb))\n",
    "    return C\n",
    "def matrix_scatterplot(M, xticklabels, yticklabels,legend_name=None, xlabel='', ylabel='', fig_width=10, fig_height=14, scale_factor=500,save_dot=None):\n",
    "    \"\"\"Plots a matrix with points as in a scatterplot. Area of points proportional to each matrix element. \n",
    "    Suitable to show sparse matrices.\n",
    "\n",
    "    Args:\n",
    "        M (np.array): a 2D array\n",
    "        xticklabels: label list\n",
    "        yticklabels: label list\n",
    "        fig_width (int): matplotlib figure width\n",
    "        fig_height (int): matplotlib figure height\n",
    "        scale_factor (float): scales the points by this value. \n",
    "    \"\"\"\n",
    "#     M = M / M.sum(axis=0)\n",
    "    M_norm = M / M.sum(axis=0)\n",
    "    Mplot = M_norm.copy()*scale_factor\n",
    "#     Mplot = Mplot / Mplot.sum(axis=0)\n",
    "    Mplot = np.flip(Mplot, axis=0)\n",
    "    yticklabels.reverse()\n",
    "    x = np.arange(0, M.shape[1], 1)\n",
    "    y = np.arange(0, M.shape[0], 1)\n",
    "    xx, yy = np.meshgrid(x, y)\n",
    "    plt.figure(figsize=(fig_width, fig_height))\n",
    "    cmap='Set1'\n",
    "    color = color_cm(cmap,M.shape[0]+M.shape[1])\n",
    "    colors = np.repeat(color[:M.shape[0]],M.shape[1])\n",
    "    plt.scatter(np.ravel(xx), np.ravel(yy), s=np.ravel(Mplot), c=colors[::-1])\n",
    "    \n",
    "\n",
    "    ax = plt.gca()\n",
    "    ax.set_xlim(np.min(x)-0.5, np.max(x)+0.5)\n",
    "    ax.set_ylim(np.min(y)-0.5, np.max(y)+0.5)\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(xticklabels, rotation=90)\n",
    "    ax.set_yticks(y)\n",
    "    ax.set_yticklabels(yticklabels, rotation=0)\n",
    "    ax.xaxis.set_ticks_position('top')\n",
    "    ax.yaxis.set_ticks_position('left')\n",
    "    ax.set_xlabel(xlabel, fontsize=12)\n",
    "    ax.set_ylabel(ylabel, fontsize=12)\n",
    "\n",
    "    ax.tick_params(color='None')\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"bottom\"].set_visible(False)\n",
    "\n",
    "    for tick in ax.get_yticklabels():\n",
    "        #tick.set_fontname(\"DejaVu Sans Mono\")\n",
    "        tick.set_fontfamily('monospace')\n",
    "        tick.set_fontsize(12)\n",
    "\n",
    "    for tick in ax.get_xticklabels():\n",
    "        tick.set_fontfamily('monospace')\n",
    "        tick.set_fontsize(12)\n",
    "\n",
    "    plt.grid(color='gray', linestyle='-', linewidth=0.5, alpha=0.4)\n",
    "    plt.box(False)\n",
    "    plt.tight_layout()\n",
    "    l1 = plt.scatter([],[], s=scale_factor/4, edgecolors='none',c='gray')\n",
    "    l2 = plt.scatter([],[], s=scale_factor*2/4, edgecolors='none',c='gray')\n",
    "    l3 = plt.scatter([],[], s=scale_factor*3/4, edgecolors='none',c='gray')\n",
    "    l4 = plt.scatter([],[], s=scale_factor, edgecolors='none',c='gray')\n",
    "\n",
    "    labels = [\"25%\", \"50%\", \"75%\", \"100%\"]\n",
    "\n",
    "    legend1 = plt.legend([l1, l2, l3, l4], labels, ncol=1, frameon=True, fontsize=30,\n",
    "    handlelength=2, loc=[1.1, 0.1], borderpad = 1.8,\n",
    "    handletextpad=1, title='matching percentage', scatterpoints = 1)\n",
    "    ax = plt.gca().add_artist(legend1)\n",
    "    l_co = []\n",
    "    for cl in color[:M.shape[0]]:\n",
    "        l_co.append(plt.scatter([],[], s=scale_factor/4, edgecolors='none',c=cl))\n",
    "\n",
    "    labels = legend_name\n",
    "\n",
    "    legend2 =plt.legend(l_co, labels, ncol=1, frameon=True, fontsize=18,\n",
    "    handlelength=2, loc=[1.1, 0.8], borderpad = 1.8,\n",
    "    handletextpad=1, title='matching percentage', scatterpoints = 1)\n",
    "    \n",
    "\n",
    "    if save_dot:\n",
    "        plt.savefig(save_dot,bbox_extra_artists=(legend1,legend2), bbox_inches='tight')\n",
    "    plt.show()\n",
    "    return\n",
    "def consistency_plots(t_type,e_type,t_name='T',e_name='E',re_order=True, scale_factor=500,xlabel=r'Transcriptomic type',\n",
    "                       ylabel=r'Electrophysiology type',save_dot=None,save_river=None,save_cf = None,save_cf_n = None):\n",
    "#     print(np.unique(e_type),np.unique(t_type))\n",
    "    C_e_types = contingency(a=e_type,\n",
    "                  b=t_type,\n",
    "                  unique_a=np.unique(e_type),\n",
    "                  unique_b=np.unique(t_type))\n",
    "    #Assign labels of clusters based on 'best match' with transcriptomic celltype label\n",
    "    row_ind,col_ind = linear_sum_assignment(-C_e_types)\n",
    "    \n",
    "    #print(np.unique(e_type),np.unique(t_type),C_e_types.shape)\n",
    "    \n",
    "    C_ordered = C_e_types[:,col_ind]\n",
    "    order_y = np.unique(t_type)[col_ind]\n",
    "\n",
    "    t_labels_matched = t_type.copy()\n",
    "    e_labels_matched = e_type.copy()\n",
    "    if re_order:\n",
    "        for name in order_y:\n",
    "            ind = t_type == name\n",
    "            t_labels_matched[ind] = name\n",
    "\n",
    "    C_TE_consistency = contingency(a=e_labels_matched,\n",
    "                    b=t_labels_matched,\n",
    "                    unique_a=np.unique(e_labels_matched),\n",
    "                    unique_b=np.unique(t_labels_matched))\n",
    "\n",
    "    ax_xticklabels = [t_name+'-{}{:>4s}'.format(x,'({:d})'.format(np.sum(t_labels_matched==x))) for x in np.unique(t_labels_matched)]\n",
    "    ax_yticklabels = [e_name+'-{} ({:d})'.format(y,np.sum(e_labels_matched==y)) for y in np.unique(e_labels_matched)]\n",
    "    legend_name= [e_name+'-{}'.format(x) for x in np.unique(e_labels_matched)]\n",
    "\n",
    "    #print(np.unique(e_labels_matched),np.unique(t_labels_matched),C_TE_consistency.shape)\n",
    "    matrix_scatterplot(M=C_TE_consistency,scale_factor=scale_factor,\n",
    "                       xticklabels=ax_xticklabels,\n",
    "                       yticklabels=ax_yticklabels,\n",
    "                       xlabel=xlabel,\n",
    "                       ylabel=ylabel,\n",
    "                       legend_name= legend_name,\n",
    "                       fig_width=8,fig_height=8,save_dot=save_dot)\n",
    "\n",
    "    ax_xticklabels = [t_name+'-{}'.format(x) for x in np.unique(t_labels_matched)]\n",
    "    ax_yticklabels = [e_name+'-{}'.format(y) for y in np.unique(e_labels_matched)]\n",
    "    df_et = pd.DataFrame(C_TE_consistency, columns = ax_xticklabels,\n",
    "                       index= ax_yticklabels)\n",
    "    plt.figure()\n",
    "    sn.heatmap(df_et)\n",
    "#     plt.title('gene clustering VS ephys cluster ')\n",
    "    if save_cf_n:\n",
    "        plt.savefig(save_cf_n, bbox_inches='tight')\n",
    "        \n",
    "    plt.figure()\n",
    "    sn.heatmap(df_et.div(df_et.sum(axis=1), axis=0))\n",
    "#     plt.title('gene clustering VS  ephys cluster ')\n",
    "    if save_cf:\n",
    "        plt.savefig(save_cf, bbox_inches='tight')\n",
    "   \n",
    "    df_river = {'e type':[],'t type':[],'et value':[]}\n",
    "    for t_cell in df_et.columns.to_list():\n",
    "        for e_cell in df_et.index.to_list():\n",
    "            df_river['e type'].append(e_cell)\n",
    "            df_river['t type'].append(t_cell)\n",
    "            df_river['et value'].append(df_et.loc[e_cell,t_cell])\n",
    "    df_river = pd.DataFrame(df_river)\n",
    "\n",
    "    all_nodes = df_river['e type'].unique().tolist() + df_river['t type'].unique().tolist()\n",
    "    source_indices = [all_nodes.index(e_type) for e_type in df_river['e type']]\n",
    "    target_indices = [all_nodes.index(t_type) for t_type in df_river['t type']]\n",
    "\n",
    "    NUM_COLORS = len(all_nodes)\n",
    "    cmap='Set1'\n",
    "    colors = color_cm(cmap,NUM_COLORS)\n",
    "\n",
    "\n",
    "    # colors = pex.colors.qualitative.D3\n",
    "\n",
    "    # node_colors_mappings = dict([(node,np.random.choice(colors)) for node in all_nodes])\n",
    "    node_colors_mappings = dict([(node,colors[idx]) for idx,node in enumerate(all_nodes)])\n",
    "    node_colors = [node_colors_mappings[node] for node in all_nodes]\n",
    "    edge_colors = [node_colors_mappings[node] for node in df_river['e type']]\n",
    "\n",
    "    fig = go.Figure(data=[go.Sankey(\n",
    "        node = dict(\n",
    "          pad = 20,\n",
    "          thickness = 20,\n",
    "          line = dict(color = \"black\", width = 1.0),\n",
    "          label =  all_nodes,\n",
    "          color =  node_colors,\n",
    "        ),\n",
    "\n",
    "        link = dict(\n",
    "          source =  source_indices,\n",
    "          target =  target_indices,\n",
    "          value =  df_river['et value'],\n",
    "          color = edge_colors,\n",
    "    ))])\n",
    "\n",
    "    fig.update_layout(title_text=\"e type to t type\",\n",
    "                      height=600,\n",
    "                      font_size=10)\n",
    "    if save_river:\n",
    "        fig.write_image(save_river)\n",
    "    fig.show()\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "stage = gene_set.obs['stage'].to_numpy()\n",
    "m_type = joint_set.obs['cluster'].to_numpy()\n",
    "name1='M'\n",
    "name2='stage'\n",
    "re_order=False\n",
    "consistency_plots(m_type,stage,t_name=name1,e_name=name2,re_order=True, scale_factor=2000,xlabel=r'Joint clustering type',\n",
    "                       ylabel=r'Stage',save_dot='./figures/dot_joint_clustering.pdf',save_river='./figures/river_joint_clustering.pdf',\n",
    "                  save_cf = './figures/joint_single_cf.pdf',save_cf_n = './figures/joint_single_cf_norm.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stage = gene_set.obs['stage'].to_numpy()\n",
    "t_type = gene_set.obs['cell type'].to_numpy()\n",
    "name1='T'\n",
    "name2='stage'\n",
    "\n",
    "consistency_plots(t_type,stage,t_name=name1,e_name=name2,re_order=True, scale_factor=2000,xlabel=r'Transcriptomic type',\n",
    "                       ylabel=r'Stage',save_dot='./figures/gene_single_dot.pdf',save_river='./figures/gene_single_river.pdf',save_cf = './figures/gene_single_cf.pdf',save_cf_n = 'gene_single_cf_norm.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stage = ephys_set.obs['stage'].to_numpy()\n",
    "t_type = ephys_set.obs['cell type'].to_numpy()\n",
    "name1='E'\n",
    "name2='stage'\n",
    "\n",
    "consistency_plots(t_type,stage,t_name=name1,e_name=name2,re_order=True, scale_factor=2000,xlabel=r'Electrophysiology type',\n",
    "                       ylabel=r'Stage',save_dot='./figures/ephys_single_dot.pdf',save_river='./figures/ephys_single_river.pdf',save_cf = './figures/ephys_single_cf.pdf',save_cf_n = 'ephys_single_cf_norm.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pseudo time analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "colors=['b','g','y','r']\n",
    "sns.displot(gene_set.obs,x='pseudo', hue='stage', kind='kde', fill=True,palette=colors ,rug=True)\n",
    "plt.savefig('./figures/Gene_pseuro_Shadow.pdf')\n",
    "plt.show()\n",
    "sns.displot(ephys_set.obs,x='pseudo', hue='stage', kind='kde', fill=True,palette=colors ,rug=True)\n",
    "plt.savefig('./figures/Ephys_pseuro_Shadow.pdf')\n",
    "plt.show()\n",
    "sns.displot(joint_set.obs,x='pseudo', hue='stage', kind='kde', fill=True,palette=colors ,rug=True)\n",
    "plt.savefig('./figures/Joint_pseuro_Shadow.pdf')\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fvalues=[]\n",
    "pvalues=[]\n",
    "label=['gene_contact','ephys','joint']\n",
    "import scipy.stats as stats\n",
    "# stats f_oneway functions takes the groups as input and returns ANOVA F and p value\n",
    "fvalue, pvalue  = stats.f_oneway(gene_set[gene_set.obs['stage']=='day12'].obs['pseudo'], \n",
    "                                gene_set[gene_set.obs['stage']=='day21'].obs['pseudo'],\n",
    "                                gene_set[gene_set.obs['stage']=='day46'].obs['pseudo'],\n",
    "                                gene_set[gene_set.obs['stage']=='day64'].obs['pseudo'])\n",
    "fvalues.append(fvalue)\n",
    "pvalues.append(pvalue)\n",
    "fvalue, pvalue = stats.f_oneway(ephys_set[ephys_set.obs['stage']=='day12'].obs['pseudo'], \n",
    "                                ephys_set[ephys_set.obs['stage']=='day21'].obs['pseudo'],\n",
    "                                ephys_set[ephys_set.obs['stage']=='day46'].obs['pseudo'],\n",
    "                                ephys_set[ephys_set.obs['stage']=='day64'].obs['pseudo'])\n",
    "fvalues.append(fvalue)\n",
    "pvalues.append(pvalue)\n",
    "fvalue, pvalue = stats.f_oneway(joint_set[joint_set.obs['stage']=='day12'].obs['pseudo'], \n",
    "                                joint_set[joint_set.obs['stage']=='day21'].obs['pseudo'],\n",
    "                                joint_set[joint_set.obs['stage']=='day46'].obs['pseudo'],\n",
    "                                joint_set[joint_set.obs['stage']=='day64'].obs['pseudo'])\n",
    "fvalues.append(fvalue)\n",
    "pvalues.append(pvalue)\n",
    "\n",
    "\n",
    "\n",
    "label=['gene_contact','ephys','joint']\n",
    "import scipy.stats as stats\n",
    "stat_test_pseudo = pd.DataFrame(index=label)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import calinski_harabasz_score\n",
    "metrc_score = []\n",
    "metrc_score.append(calinski_harabasz_score(gene_set.X, gene_set.obs['cell type']))\n",
    "metrc_score.append(calinski_harabasz_score(ephys_set.X, ephys_set.obs['cell type']))\n",
    "metrc_score.append(calinski_harabasz_score(joint_set.X, joint_set.obs['cluster']))\n",
    "stat_test_pseudo['calinski_harabasz_score']=metrc_score\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from sklearn.metrics import silhouette_score\n",
    "from itertools import combinations\n",
    "comb = list(combinations([0, 1, 2, 3], 2))\n",
    "p_matrix_ttest=pd.DataFrame(data=np.ones([4,4]), index=stages, columns=stages)\n",
    "# p_matrix_ttest[:] = np.nan\n",
    "for co in comb:\n",
    "    idx = gene_set.obs['stage'].isin([stages[co[0]],stages[co[1]]])\n",
    "    p_matrix_ttest.iloc[[co[1]],[co[0]]] = silhouette_score(gene_set[idx].obsm['X_umap'],\n",
    "                    gene_set[idx].obs['stage'].values)\n",
    "    p_matrix_ttest.iloc[[co[0]],[co[1]]] = p_matrix_ttest.iloc[[co[1]],[co[0]]]\n",
    "\n",
    "print(p_matrix_ttest)\n",
    "fig,ax=plt.subplots()\n",
    "plt.title('contact gene clustering score')\n",
    "sns.heatmap(p_matrix_ttest,annot=True,ax=ax,cmap='YlGnBu', vmin=0, vmax=1, linewidths=0.7)\n",
    "plt.savefig('./figures/silhouette contact gene clustering score.pdf')\n",
    "\n",
    "\n",
    "comb = list(combinations([0, 1, 2, 3], 2))\n",
    "p_matrix_ttest=pd.DataFrame(data=np.ones([4,4]), index=stages, columns=stages)\n",
    "# p_matrix_ttest[:] = np.nan\n",
    "for co in comb:\n",
    "    idx = ephys_set.obs['stage'].isin([stages[co[0]],stages[co[1]]])\n",
    "    p_matrix_ttest.iloc[[co[1]],[co[0]]] = silhouette_score(ephys_set[idx].X,\n",
    "                    ephys_set[idx].obs['stage'].values)\n",
    "    p_matrix_ttest.iloc[[co[0]],[co[1]]] = p_matrix_ttest.iloc[[co[1]],[co[0]]]\n",
    "\n",
    "print(p_matrix_ttest)\n",
    "fig,ax=plt.subplots()\n",
    "plt.title('contact ephys clustering score')\n",
    "sns.heatmap(p_matrix_ttest,annot=True,ax=ax,cmap='YlGnBu', vmin=0, vmax=1, linewidths=0.7)\n",
    "plt.savefig('./figures/silhouette contact ephys clustering score.pdf')\n",
    "\n",
    "\n",
    "comb = list(combinations([0, 1, 2, 3], 2))\n",
    "p_matrix_ttest=pd.DataFrame(data=np.ones([4,4]), index=stages, columns=stages)\n",
    "# p_matrix_ttest[:] = np.nan\n",
    "for co in comb:\n",
    "    idx = joint_set.obs['stage'].isin([stages[co[0]],stages[co[1]]])\n",
    "    p_matrix_ttest.iloc[[co[1]],[co[0]]] = silhouette_score(joint_set[idx].X,\n",
    "                    joint_set[idx].obs['stage'].values)\n",
    "    p_matrix_ttest.iloc[[co[0]],[co[1]]] = p_matrix_ttest.iloc[[co[1]],[co[0]]]\n",
    "\n",
    "print(p_matrix_ttest)\n",
    "fig,ax=plt.subplots()\n",
    "plt.title('joint clustering score')\n",
    "sns.heatmap(p_matrix_ttest,annot=True,ax=ax,cmap='YlGnBu', vmin=0, vmax=1, linewidths=0.7)\n",
    "plt.savefig('./figures/silhouette joint clustering score.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# neuron analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_cells = sc.read_h5ad('./SourceData/CulturedNeuronGeneAll.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi_save=300,figsize=[4.0, 4.0])\n",
    "sc.pl.umap(overall_cells, color=['cell type'],palette = 'rainbow', legend_loc='right margin',save='umap_neuron_all_type.png')\n",
    "sc.pl.dotplot(overall_cells, ['SLC17A6','GAD1','GFAP','LUM'], groupby=f'cell type',swap_axes=True,\n",
    "              save='All_neuron_non_neuron_dotplot_specific_marker_major.png')\n",
    "marker_names=pd.DataFrame(overall_cells.uns['rank_genes_groups']['names']).head(5).to_numpy().ravel(order='F')\n",
    "_,idx = list(np.unique(marker_names,return_index=True))\n",
    "markers = marker_names[np.sort(idx)]\n",
    "sc.pl.heatmap(overall_cells, markers, log=True,groupby='cell type',cmap='Reds'\n",
    "              ,swap_axes=True,save='All_neuron_non_neuron_heat_map_major.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neuron_expression = sc.read_h5ad('./SourceData/CulturedNeuronGenesRecord.h5ad')\n",
    "neuron_ephys = sc.read_h5ad('./SourceData/CulturedNeuronEphysRecord.h5ad')\n",
    "markers = ['SLC17A6','SATB2','NREP','GAD1','BHLHE22','PHOX2A']\n",
    "\n",
    "sc.pl.heatmap(neuron_expression, markers, log=False,groupby=f'cell type',cmap='Reds'\n",
    "              ,swap_axes=False,save='located_neuron_gene_heatmap_majortype.pdf')\n",
    "sc.pl.heatmap(neuron_ephys, ['Fano_factor','firing_rate','recovery_slope','repolarization_slope','peak_trough_ratio','peak_to_valley'],\n",
    "              log=False, swap_axes=False, groupby=f'cell type', cmap='bwr', \n",
    "              show=True,save='located_neuron_ephys_heatmap_majortype.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "metadata": {
     "collapsed": false
    },
    "source": []
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
